### PR TITLE
Fix #28011 - remove page number from query on layered navigation swatch filter

### LINKED ISSUE #27989: Elasticsearch attribute filter with pagination issue
**Elasticsearch attribute filter with pagination issue**

### Preconditions (*)
1. Magento 2.4-develop;
2. Sample Data;
2. Elasticsearch 6 +

### Steps to reproduce:

1. Go to Storefront;
2. Go to Men->Tops->Hoodies & Sweatshirts category (for example);
3. Go to the second **2** page;
4. On filter panel choose **black** from **Color** (for example);

**Actual result:**

:heavy_multiplication_x: After applying the filter, the products are not displayed

Peek 2020-04-30 14-42

5. :heavy_check_mark:  When deleting `p=2` from URL path, products are displayed

Peek 2020-04-30 14-51

**Expected result:**

1. it should show the filtered products.
2.  if I am on the 2th page and  filtered any attribute then it should remove the pagination value from the URL or it should be managed with the product item count
3. after removing the pagination value from the URL  it's showing the 6 number of products

### LINKED ISSUE #28011: ElasticSearch When applying the extra filter to the product listing, Magento keeps the same page number
### Preconditions (*)

1. Magento 2.3.3
2. PHP 7.2
3. elasticsearch 6+

### Steps to reproduce (*)

1. Enter listing with n products,
2. Move to the last page of results
3. Select filter in the sidebar that filters out most of the products

### Expected result (*)

1.I'm redirected either to last page of results or first page of the listing (so I see any products)

### Actual result (*)

1. I'm still at the last page of results before adding the filter, so the message says that there are no search results for my query

**Here is the video**: [Desktop-2020-4-28-15-16-12.webm.zip](https://github.com/magento/magento2/files/4544871/Desktop-2020-4-28-15-16-12.webm.zip)

### PR DESCRIPTION
**Pull request by Bartlomiejsz**
### Description (*)

This change makes swatch layered navigation filter consistent with standard filters, by removing `p` query parameter from filter url in same way as in https://github.com/magento/magento2/blob/2.4-develop/app/code/Magento/Catalog/Model/Layer/Filter/Item.php
Fixes #28011 

### Fixed Issues (if relevant)

1. magento/magento2#28011: ElasticSearch When applying the extra filter to the product listing, Magento keeps the same page number

### Manual testing scenarios (*)

1. Have at least one swatch filter on layered navigation and multiple pages of products
2. Go to last page on PLP
3. Use swatch attribute as filter
4. `p` parameter should be removed from url and user should see first page, same as when non-swatch attribute is used as filter

### Contribution checklist (*)
 - [x] Pull request has a meaningful description of its purpose
 - [x] All commits are accompanied by meaningful commit messages
 - [x] All new or changed code is covered with unit/integration tests (if applicable)
 - [x] All automated tests passed successfully (all builds are green)

### CODE REVIEW

File: app/code/Magento/Swatches/Block/LayeredNavigation/RenderLayered.php [hunk 1]
```diff
@@ -192,7 +216,7 @@ protected function getOptionViewData(FilterItem $filterItem, Option $swatchOptio
      */
     protected function isOptionVisible(FilterItem $filterItem)
     {
-        return $this->isOptionDisabled($filterItem) && $this->isShowEmptyResults() ? false : true;
+        return !($this->isOptionDisabled($filterItem) && $this->isShowEmptyResults());
```
Comment (lbajsarowicz) [inline app/code/Magento/Swatches/Block/LayeredNavigation/RenderLayered.php:239]: Nice catch! :+1:

**Review by lbajsarowicz — DISMISSED** [review]
As it's an adaptation of another bugfix (made to another part of the system), I am approving that.
Thanks for cleaning up the code style there.

**Comment by magento-engcom-team** [comment]
Hi @lbajsarowicz, thank you for the review.
[ENGCOM-7493](https://jira.corp.magento.com/browse/ENGCOM-7493) has been created to process this Pull Request
:eight_spoked_asterisk: @lbajsarowicz, could you please add one of the following labels to the Pull Request?

| Label | Description |
|--------------------------------|-----------------------------------------------------------------|
| ```Auto-Tests: Covered``` | All changes in Pull Request is covered by auto-tests |
| ```Auto-Tests: Not Covered``` | Changes in Pull Request requires coverage by auto-tests |
| ```Auto-Tests: Not Required``` | Changes in Pull Request does not require coverage by auto-tests |

**Review by ihor-sviziev — CHANGES_REQUESTED** [review]
Hi @Bartlomiejsz,
Could you fix the failing static tests and cover your change with any type of test? I see you already adjusted unit tests, but I can't see any change related to the bug itself.

**Comment by Bartlomiejsz** [comment]
Hi @ihor-sviziev, indeed I only did minor adjustments to unit test because otherwise it would fail due to new dependency injected. I didn't add new unit test since this whole test file will either need to be reworked or removed for phpunit 9 because it tests mock and not real class which I think won't work anymore.
I will try to cover this change with functional test

File: app/code/Magento/Swatches/Test/Mftf/Test/StorefrontRedirectToFirstPageOnFilteringBySwatchTest.xml [hunk 2]
```diff
@@ -0,0 +1,103 @@
...
+        <actionGroup ref="AdminOpenProductIndexPageActionGroup" stepKey="goToProductsGridPage3"/>
+        <actionGroup ref="OpenEditProductOnBackendActionGroup" stepKey="goToProduct3EditPage">
+            <argument name="product" value="$$createSimpleProduct3$$"/>
+        </actionGroup>
+        <selectOption selector="{{AdminProductAttributesSection.attributeDropdownByCode(textSwatchAttribute.attribute_code)}}" userInput="textSwatchOption2" stepKey="selectProduct3AttributeOption"/>
+        <actionGroup ref="SaveProductFormActionGroup" stepKey="saveProduct3"/>
+
+        <magentoCLI command="cron:run --group=index" stepKey="runCronIndexer"/>
+
+        <amOnPage url="$$createCategory.name$$.html" stepKey="amOnCategoryPage"/>
```
Comment (lbajsarowicz) [inline app/code/Magento/Swatches/Test/Mftf/Test/StorefrontRedirectToFirstPageOnFilteringBySwatchTest.xml]: There's an Action Group that uses `url_key`, not `name` of category.
I can look for it tomorrow.

File: app/code/Magento/Swatches/Test/Mftf/Test/StorefrontRedirectToFirstPageOnFilteringBySwatchTest.xml [hunk 3]
```diff
@@ -0,0 +1,103 @@
...
+        <actionGroup ref="SaveProductFormActionGroup" stepKey="saveProduct2"/>
+
+        <actionGroup ref="AdminOpenProductIndexPageActionGroup" stepKey="goToProductsGridPage3"/>
+        <actionGroup ref="OpenEditProductOnBackendActionGroup" stepKey="goToProduct3EditPage">
+            <argument name="product" value="$$createSimpleProduct3$$"/>
+        </actionGroup>
+        <selectOption selector="{{AdminProductAttributesSection.attributeDropdownByCode(textSwatchAttribute.attribute_code)}}" userInput="textSwatchOption2" stepKey="selectProduct3AttributeOption"/>
+        <actionGroup ref="SaveProductFormActionGroup" stepKey="saveProduct3"/>
+
+        <magentoCLI command="cron:run --group=index" stepKey="runCronIndexer"/>
```
Comment (lbajsarowicz) [inline app/code/Magento/Swatches/Test/Mftf/Test/StorefrontRedirectToFirstPageOnFilteringBySwatchTest.xml]:
```suggestion
        <magentoCron groups="index" stepKey="runCronIndexer"/>
```

**Review by lbajsarowicz — CHANGES_REQUESTED** [review]
Actually, I really like the way that you prepared the test. It's really nice. There are only minor issues, let me send you the recommended Action Group for opening category tomorrow.

**Comment by Bartlomiejsz** [comment]
Hi @lbajsarowicz, I believe I found actionGroup that you mentioned. I updated PR with changes from review, also I extracted one part to action group, added fix for issue caused by other tests not being isolated and added `block_html` and `collections` caches to be cleaned, since I noticed those cache types are also invalidated when eav attribute is created

File: app/code/Magento/Catalog/Test/Mftf/ActionGroup/AssertStorefrontCategoryCurrentPageIsFirstActionGroup.xml [hunk 4]
```diff
@@ -0,0 +1,17 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<!--
+ /**
+  * Copyright © Magento, Inc. All rights reserved.
+  * See COPYING.txt for license details.
+  */
+-->
+<actionGroups xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
+              xsi:noNamespaceSchemaLocation="urn:magento:mftf:Test/etc/actionGroupSchema.xsd">
+    <actionGroup name="AssertStorefrontCategoryCurrentPageIsFirstActionGroup">
```
Comment (lbajsarowicz) [inline app/code/Magento/Catalog/Test/Mftf/ActionGroup/AssertStorefrontCategoryCurrentPageIsFirstActionGroup.xml]: How about doing that more universal by changing to StorefrontAssertCategoryCurrentPageIsNthActionGroup and pass expectedPage as an argument?

File: app/code/Magento/Catalog/Test/Mftf/ActionGroup/StorefrontNavigateCategoryNextPageActionGroup.xml [hunk 5]
```diff
@@ -0,0 +1,15 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<!--
+/**
+ * Copyright © Magento, Inc. All rights reserved.
+ * See COPYING.txt for license details.
+-->
+
+<actionGroups xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:noNamespaceSchemaLocation="urn:magento:mftf:Test/etc/actionGroupSchema.xsd">
+    <!-- Navigates storefront category next page from toolbar -->
```
Comment (lbajsarowicz) [inline app/code/Magento/Catalog/Test/Mftf/ActionGroup/StorefrontNavigateCategoryNextPageActionGroup.xml]: There is <annotation> node for descriptions :-)

File: app/code/Magento/Catalog/Test/Mftf/Section/StorefrontCategoryBottomToolbarSection.xml [hunk 6]
```diff
@@ -12,6 +12,6 @@
         <element name="previousPage" type="button" selector=".//*[@class='toolbar toolbar-products'][2]//a[contains(@class, 'previous')]" timeout="30"/>
         <element name="pageNumber" type="text" selector="//*[@class='toolbar toolbar-products'][2]//a[contains(@class, 'page')]//span[2][contains(text() ,'{{var1}}')]" parameterized="true"/>
         <element name="perPage" type="select" selector="//*[@class='toolbar toolbar-products'][2]//select[@id='limiter']"/>
-        <element name="currentPage" type="text" selector=".products.wrapper + .toolbar-products .pages .current span:nth-of-type(2)"/>
+        <element name="currentPage" type="text" selector=".//*[@class='toolbar toolbar-products'][2]//li[contains(@class, 'current')]//span[2]" timeout="30"/>
```
Comment (lbajsarowicz) [inline app/code/Magento/Catalog/Test/Mftf/Section/StorefrontCategoryBottomToolbarSection.xml:15]: Is there any other way - without using [n] to point the bottom pagination? Some can just change the top one and... The selector won't work for bottom one.

File: app/code/Magento/Swatches/Test/Mftf/Test/StorefrontRedirectToFirstPageOnFilteringBySwatchTest.xml [hunk 7]
```diff
@@ -0,0 +1,102 @@
...
+            </createData>
+
+            <magentoCLI command="config:set catalog/frontend/grid_per_page 1" stepKey="setOneProductPerPage"/>
+            <magentoCLI command="config:set catalog/frontend/grid_per_page_values 1" stepKey="setGridPerPage"/>
+
+            <actionGroup ref="AdminLoginActionGroup" stepKey="loginAsAdmin"/>
+            <actionGroup ref="AddTextSwatchToProductActionGroup" stepKey="addSwatchAttribute">
+                <argument name="usedInLayeredNavigation" value="1"/>
+            </actionGroup>
+            <magentoCLI command="cache:clean block_html collections config full_page" stepKey="cleanInvalidatedCaches"/>
```
Comment (lbajsarowicz) [inline app/code/Magento/Swatches/Test/Mftf/Test/StorefrontRedirectToFirstPageOnFilteringBySwatchTest.xml]: magentoCron groups=index should be enough, as it cleans cache selectively for the changed products.

**Review by lbajsarowicz — CHANGES_REQUESTED** [review]
Now it's only about minor changes 👏

File: app/code/Magento/Catalog/Test/Mftf/Section/StorefrontCategoryBottomToolbarSection.xml (hunk 6 above)
Comment (Bartlomiejsz) [reply app/code/Magento/Catalog/Test/Mftf/Section/StorefrontCategoryBottomToolbarSection.xml:15]: I don't think there is any reliable way to do that. Changing it into `.products.wrapper + .toolbar-products` as it was before has even more chance to fail, as it's relatively easy to insert something between products and bottom toolbar, but moving or removing one of toolbars is one of things that are currently impossible - https://github.com/magento/magento2/issues/7253 🙂

**Review by lbajsarowicz — DISMISSED** [review]
I have no more comments :) Great job, thank you.

**Comment by magento-engcom-team** [comment]
Hi @lbajsarowicz, thank you for the review.
[ENGCOM-7493](https://jira.corp.magento.com/browse/ENGCOM-7493) has been created to process this Pull Request

**Review by ihor-sviziev — DISMISSED** [review]
Good job 👍

**Comment by magento-engcom-team** [comment]
Hi @ihor-sviziev, thank you for the review.
[ENGCOM-7493](https://jira.corp.magento.com/browse/ENGCOM-7493) has been created to process this Pull Request

**Comment by engcom-Alfa** [comment]
:heavy_check_mark: QA Passed

_**Preconditions:**_

Sample Data;
Elasticsearch 6+;

**Steps to Reproduce:**
1. Go to Storefront;
2. Go to Men->Tops->Hoodies & Sweatshirts category (for example);
3. Go to the second **2** page;
4. On filter panel choose **black** from **Color** (for example);

**Before:**

:heavy_multiplication_x: After applying the filter, the products are not displayed

Peek 2020-04-30 14-42

**After:** we see filtered products on the first page

Peek 2020-05-04 13-47

_Note:_ Tested with visual and text swatches type

**Review by slavvka — CHANGES_REQUESTED** [review]
@Bartlomiejsz please resolve the conflicts and re-run the tests after the upgrade to PHP 7.4

**Comment by Bartlomiejsz** [comment]
Hi @slavvka, done

**Review by ihor-sviziev — APPROVED** [review]

**Comment by magento-engcom-team** [comment]
Hi @ihor-sviziev, thank you for the review.
[ENGCOM-7493](https://jira.corp.magento.com/browse/ENGCOM-7493) has been created to process this Pull Request

**Comment by engcom-Alfa** [comment]
:heavy_check_mark: QA Passed

**Comment by engcom-Charlie** [comment]
Failed functional test not related to changes from this PR.

**Pull request merged**